fix writing canvas past bounds

labseven opened the pull request:
It works nicely.
Also fixed not writing last line on the right and bottom.

fixes #31

newsch committed:
Add util header with eprintf

MatthewBeaudouinLafon committed:
first pass free line, featuring bad diagonals

MatthewBeaudouinLafon committed:
added last cursor position, doesn't work too well

MatthewBeaudouinLafon committed:
gave up on cleaning diagonal lines due to symmetry

MatthewBeaudouinLafon committed:
clean up

Commit:
fix canvas bounds issues

newsch commented on src/frontend.c:
```diff
@@ -115,12 +115,26 @@ void front_setcharcursor(char ch) {
 void redraw_canvas_win() {
-  for (int x = 0; x < view_max_x; x++) {
-    for (int y = 0; y < view_max_y; y++) {
+  int max_x = view_max_x;
```
These sections could be commented more.

newsch commented on src/cursor.c:
```diff
@@ -23,6 +23,9 @@ void cursor_move_up(Cursor *cursor, View *view) {
 }
 
 void cursor_move_down(Cursor *cursor, View *view) {
+  if (cursor->y == view->canvas->num_rows - view->y - 1){
```
I can't help but feel that checking for `==` instead of `>=`/`<=` will come back to bite us at some point, but it seems fine for now.

newsch committed:
Comment new redraw_canvas_win steps

newsch committed:
Draw status window before first keypress

newsch committed:
Remove old call to update_screen_size

The function doesn't accept any parameters, but there weren't any
compiler warnings. I just learned that `function()` is different
from `function(void)` - the former is a prototype, not a function
that accepts zero arguments.

> Welcome to the C language, this is part of the legacy standard.

https://stackoverflow.com/q/12643202

newsch commented:
I get a segfault when the canvas is too short, but not when it is too thin.

newsch commented on src/frontend.c:
```diff
@@ -115,12 +115,29 @@ void front_setcharcursor(char ch) {
+
+  // draw fill in rest of window
+  for (int x = max_x; x < view_max_x; x++) {
+    for (int y = 0; y < view_max_y; y++) {
```
Should this be:
```suggestion
    for (int y = max_y; y < view_max_y; y++) {
```

newsch commented on src/frontend.c:
```diff
@@ -115,12 +115,29 @@ void front_setcharcursor(char ch) {
+
+  // draw canvas onto window
+  for (int x = 0; x <= max_x; x++) {
+    for (int y = 0; y <= max_y; y++) {
```
Starting at 0 and drawing at index +1 feels weird to me. Is this to get around the border width?

newsch replied:
Or is this compensating for the window border?

labseven commented:
> I get a segfault when the canvas is too short, but not when it is too thin.

Fixed by messing with some magic numbers in `cursor` and `frontend`

newsch committed:
Redirect stderr

newsch committed:
Add debug log macro

newsch committed:
Add define to enable logfile

newsch commented:
I took a fresh look at it today. Starting with a view at (0, 0) and a canvas of size 10x10 I only get a usable space of 9x9:

```
┌──────────────────
│123456789XXXXXXXXX
│2        XXXXXXXXX
│3        XXXXXXXXX
│4        XXXXXXXXX
│5        XXXXXXXXX
│6        XXXXXXXXX
│7        XXXXXXXXX
│8        XXXXXXXXX
│9        XXXXXXXXX
│XXXXXXXXXXXXXXXXXX
│XXXXXXXXXXXXXXXXXX
│XXXXXXXXXXXXXXXXXX
│XXXXXXXXXXXXXXXXXX
```

newsch commented:
Looks like it's overwriting the last row/column:

```
┌────────────
│0        XXX
│         XXX
│         XXX
│         XXX
│         XXX
│         XXX
│         XXX
│         XXX
│        8XXX
│XXXXXXXXXXXX
│XXXXXXXXXXXX
│XXXXXXXXXXXX
```

To replicate:
```patch
diff --git a/src/frontend.c b/src/frontend.c
index 17fc043..e592c2a 100644
--- a/src/frontend.c
+++ b/src/frontend.c
@@ -51,9 +51,12 @@ int main(int argc, char *argv[]) {
   status_win = create_status_win();
 
   cursor = cursor_new();
-  Canvas *canvas = canvas_new_blank(1000, 1000);
+  Canvas *canvas = canvas_new_blank(10, 10);
+  canvas_scharyx(canvas, 0, 0, '0');
+  canvas_scharyx(canvas, 8, 8, '8');
+  canvas_scharyx(canvas, 9, 9, '9');
 
-  view = view_new_startpos(canvas, 300, 300);
+  view = view_new_startpos(canvas, 0, 0);
 
   // Enable keyboard mapping
   keypad(canvas_win, TRUE);
@@ -71,7 +74,7 @@ int main(int argc, char *argv[]) {
   State new_state = {
       .ch_in = 0,
       .cursor = cursor,
-      .current_mode = MODE_INSERT,
+      .current_mode = MODE_PAN,
       .last_arrow_direction = KEY_RIGHT,
       .last_canvas_mode = MODE_INSERT,
       .view = view,
```

labseven commented:
Yes I know my off-by-ones are messy. I'll clean it up today.

newsch commented:
I got it working in 5253be5 but it doesn't draw to the edge of the screen anymore. I think I have a fix.

newsch commented:
Another bug I found was panning the screen past the end of the canvas overwrites the border.

labseven commented:
Ok, I cleaned it up and added bounds checking

labseven commented:
I fixed the "panning the screen past the end of the canvas overwrites the border." by checking for panning bounds. It will not let you pan past the edge anymore.

newsch commented:
Looks good, but now the cursor can move to the bottom and right border again.
What are `view_max_x` and `view_max_y` supposed to represent?

labseven commented:
darn it. Let me dig deeper

newsch reviewed (approved):
Works for me. Do you want to rebase before merging it in?

newsch commented:
Looks like that had some duplicates of `free-line`. I don't think that will cause us any issues though.
